[TESTING.md] Enhance TESTING docs (#2611)
exercism/python: docs/TOOLS.md

  - [Conda](#conda)

- [Editors and IDEs](#visual-studio-code)
  - [VS Code](#visual-studio-code)
    - [Selecting an interpreter](#selecting-the-interpreter)
    - [Other features](#other-features-1)
  - [Pycharm](#pycharm)
    - [Selecting an interpreter](#selecting-the-interpreter-2)
    - [Other features](#other-features-2)
  - [Spyder](#spyder-ide)
    - [Selecting an interpreter](#selecting-the-interpreter-3)
    - [Other features](#other-features-3)

---

## Environments

Python environments are like separate Python installations, they can cleanup your workflow and projects by separating the packages you install. Making sure that you don't get bugs generated by something that you imported in another project. 

There are two major *virtual environments* in use today, `virtualenv` and `conda`. Both of which are generally easy to install.

### Venv

Also known as `virtualenv`, *venv* is a light-weight solution to virtual environments. It creates a separate Python binary for every virtual environment and stores that inside your directory.

Installing `venv` is easy using `pip`:

```bash
$ python3 -m pip install virtualenv
Successfully installed virtualenv-20.8.1

> ```

Reviewer: `venv` is now part of the standard lib:  https://docs.python.org/3/library/venv.html, so you no longer have to install `virtualenv`, you can just invoke it.  😄     Also ... Real Python has a good article on `venv` for reference.  https://realpython.com/python-virtual-environments-a-primer/.

Reviewer: [`virtualenvironmentwrapper`](https://virtualenvwrapper.readthedocs.io/en/latest/) is  very handy for management, but might also be a bit much for anyone just starting with envs.

Author: I'll remove the installing `venv` part, then! 

I am worried that realpython.com might be unusable for some students, since they now allow up to a max amount of unique visits a day. I don't think that people should see it as us trying to make them buy a realpython membership if they get hit by a paywall.

I'll add a small piece about `virtualenvironmentwrapper` at the end of the `venv` part, a bit like the ones called `other features` at the end of an IDE or Editor docs.

Reviewer: Totally agree on the RealPython thing, btw.  I am (selfishly) sad that they are metering and paywalling now.

# Tools

> A list of tools, IDEs and editors that can help you write and debug your *python* code.

Author: I am still not sure if python is written with a capital letter or not. Will look at this later again.

> A list of tools, IDEs and editors that can help you write and debug your *python* code.

Reviewer 2: If I recall correctly, the language spec is referred to as Python while the interpreter/implementation is python. Not sure if that helps in the least :D

> A list of tools, IDEs and editors that can help you write and debug your *python* code.

Author: Yep, looked it up. You are correct.

> *Disclaimer: This is a collection of tools that are popular in our community. We do not have any financial affiliation with any of the tools listed below. We think these tools do their job and there are most certainly other tools that can do those jobs as well.*

Author: "Get the job done" implies that "the tools" get one job done. "Get their job done" implies that "the tools" get the jobs done that they're designed for.

  - [Virtualenv](#venv)
  - [Conda](#conda)

- [Tools](#tools)
  - [Environments](#environments)
    - [Venv](#venv)
      - [Creating your virtual environment](#creating-your-virtual-environment)
      - [Activating your virtual environment](#activating-your-virtual-environment)
      - [Virtual Environment wrapper](#virtual-environment-wrapper)
    - [Conda](#conda)
      - [Activating your conda environment](#activating-your-conda-environment)
  - [Editors and IDEs](#editors-and-ides)
    - [Visual Studio Code](#visual-studio-code)
      - [Python for VS Code](#python-for-vs-code)
        - [Selecting the interpreter](#selecting-the-interpreter)
        - [Other features](#other-features)
    - [PyCharm](#pycharm)
      - [Selecting the interpreter](#selecting-the-interpreter-1)
      - [Other features](#other-features-1)
    - [Spyder IDE](#spyder-ide)
      - [Selecting the interpreter](#selecting-the-interpreter-2)
      - [Other features](#other-features-2)
    - [Sublime text](#sublime-text)
      - [Notable extensions](#notable-extensions)
    - [JupyterLab](#jupyterlab)
      - [Notable extensions](#notable-extensions-1)

---

## Environments

> Python environments are like separate python installations: they can organize your workflow and projects by keeping the packages you install inside that project. This helps avoid bugs generated by something that you imported in another project.

Author: ```suggestion
Python environments are like separate Python installations: they can organize your workflow and projects by keeping the packages you install inside that project. This helps avoid bugs generated by something that you imported in another project.
```

> There are two major *virtual environments* in use today, `virtualenv` and `conda`. Both of which are generally easy to install.

Author: Incorrect sentence.

> There are two major *virtual environments* in use today, `virtualenv` and `conda`. Both of which are generally easy to install.

Reviewer 2: Woops. I was going for, "Both are easy to install."

```bash
$ python3 -m virtualenv {name_of_virtualenv}
created virtual environment ... in 8568ms
```

#### Activating your virtual environment

To activate the virtual environment on Windows, run the following command:

```powershell
PS> .\{name_of_virtual_env}\Scripts\activate
(venv) PS> _
```

To activate your virtual environment on Linux, or MacOS X, run the following command:

```bash
$ source {name_of_virtual_env}/bin/activate
(venv) $ _
```

From this terminal you can now run `pip` commands. All of the packages installed in the environment will go to `{name_of_virtual_env}/Lib`.

#### Virtual Environment wrapper

Reference: [Wiki](https://virtualenvwrapper.readthedocs.io/en/latest/)

The `virtualenvwrapper` module manages all your virtual environments in one place. You can create, copy, delete and switch between environments using the tools the module provides. It also allows for extensions so you can add tools you'd like. You can create your own extension using the tutorial found [here](https://virtualenvwrapper.readthedocs.io/en/latest/plugins.html#plugins).

### Conda

> *Latest download can be found [here](https://www.anaconda.com/products/individual)*

Author: This should be a short note, not a sentence.

> Packaged with *Anaconda*, `conda` environments are similar to [venv environments](#venv) with the key difference being that `conda` can support the `R programming language`. Anaconda is most commonly used by researchers and data scientists, because of its set of tools and programs tailored to them. The [Spyder IDE](#spider-ide) is one of the tools that comes with Anaconda.

Author: There is a spelling mistake, and I dont think this makes much of a difference.

> Packaged with *Anaconda*, `conda` environments are similar to [venv environments](#venv) with the key difference being that `conda` can support the `R programming language`. Anaconda is most commonly used by researchers and data scientists, because of its set of tools and programs tailored to them. The [Spyder IDE](#spider-ide) is one of the tools that comes with Anaconda.

Reviewer 2: Corrected

To create a new `conda` environment, go to the *Anaconda Navigator*. Click on `environments` and then press the `Create` button. Fill in a name for your `conda` environment, select Python `3.8 or above` and click `Create`:

![Creating New Conda Environment](https://raw.githubusercontent.com/exercism/python/main/docs/img/Anaconda-Conda-New.png)

#### Activating your conda environment

Activating your `conda` environment is easy, just click the `►` button next to your *environment*, then press `Open Terminal`. This should open a new terminal with an interface for your `conda` environment inside.

From here you can run regular `pip` commands and other modules that you have installed inside your environment. All libraries will automatically be installed inside the `conda` environment.

---

## Editors and IDEs

### Visual Studio Code

> Visual studio code (VS Code) is a code editor created by Microsoft. It is not specialized to work for a specific programming language, but to be an editor that can do everything. You can extend the editor using extensions, but it comes with some great extensions as well.

Author: I guess it is not really "already loaded", more like 'already installed"?

> Visual studio code (VS Code) is a code editor created by Microsoft. It is not specialized to work for a specific programming language, but to be an editor that can do everything. You can extend the editor using extensions, but it comes with some great extensions as well.

Author: ```suggestion
Visual studio code (VS Code) is a code editor created by Microsoft. It is a general-purpose editor and not designed specifically for any one language. You can extend the editor using extensions; it comes with some great extensions already installed.
```

#### Python for VS Code

Extension: _Extension-id: ms-python.python_

![Python Extension Header on VS Code](https://raw.githubusercontent.com/exercism/python/main/docs/img/VSCode-EXT-Python-Header.png)

The Python extension from Microsoft is extremely useful because of its range of features. Notably it supports testing and has a testing explorer! It has many other features that you can view on [its homepage](https://marketplace.visualstudio.com/items?itemName=ms-python.python).

##### Selecting the interpreter

> The Python extensions supports the switching between multiple `interpreters`, this way you can use different Python environments for different projects. This is also useful for when you are using `venv` or `conda`, which you find more about [here](#environments).

Reviewer 2: ```suggestion
The Python extensions supports the switching between multiple `interpreters`. This allows you to use different Python environments for different projects. This is also useful for when you are using `venv` or `conda`, which you can ind more about [here](#environments).
```

Author: Spelling mistake.

Reviewer 2: Corrected

Click on the "Select interpreter" button in the lower left-hand of your window, another window should pop up where you can select the interpreter you want to use.

![Interpreter selection PT2](https://raw.githubusercontent.com/exercism/python/main/docs/img/VSCode-EXT-Python-SelectInterpreter-2.png)

##### Other features

> The Python plugin also comes with some other features that can help you debug and improve your python code, here are some of those tools.

Reviewer 2: ```suggestion
The Python plugin also comes with some other features that can help you debug and improve your python code. Below are some of those tools.
```

> [Test discovery tool](https://code.visualstudio.com/docs/python/testing#_configure-tests) - A tool that generates a tree containing all the *Pytest* and *Unittest* inside a directory. It will give you an easier and more readable interface for *Pytest*.

Reviewer 2: Pytest and Unittest ... files? Tests? Classes?

Author: Will change them to plural.

> [Test discovery tool](https://code.visualstudio.com/docs/python/testing#_configure-tests) - A tool that generates a tree containing all the *Pytest* and *Unittest* inside a directory. It will give you an easier and more readable interface for *Pytest*.

Reviewer 2: ```suggestion
[Test discovery tool](https://code.visualstudio.com/docs/python/testing#_configure-tests) - A tool that generates a tree containing all the *Pytest* and *Unittest* inside a directory. This is an easy to use and readable interface for *Pytest*.
```

> [Debugging tool](https://code.visualstudio.com/docs/python/testing#_configure-tests) - This tool will help you debug your code, it allows you to set a `breakpoint` on a line. The debugging tool then allows you to view all *private* and *global* variables at that point in your program.

Reviewer 2: ```suggestion
[Debugging tool](https://code.visualstudio.com/docs/python/testing#_configure-tests) - This tool will help you debug your code. You can set a `breakpoint` on a line. The debugging tool then allows you to view all *private* and *global* variables at that point in your program.
```

> [Linting](https://code.visualstudio.com/docs/python/testing#_configure-tests) - Linting looks out for simple mistakes in your code and notifies you when you write bad-practice code according to PEP. Exercism currently focusses on the correct use of [PEP 8](https://www.python.org/dev/peps/pep-0008/).

Reviewer 2: ```suggestion
[Linting](https://code.visualstudio.com/docs/python/testing#_configure-tests) - Linting looks out for simple mistakes in your code and notifies you when you write code which does not follow best practices according to PEP. Exercism currently focuses on following [PEP 8](https://www.python.org/dev/peps/pep-0008/).
```

> [Formatting](https://code.visualstudio.com/docs/python/editing#_formatting) - This tool automatically formats your code to look a certain way, we recommend `autopep8` because it adheres to [PEP 8](https://www.python.org/dev/peps/pep-0008/), other formatters are supported.

Reviewer 2: ```suggestion
[Formatting](https://code.visualstudio.com/docs/python/editing#_formatting) - This tool automatically formats your code to look a certain way. We recommend `autopep8` because it adheres to [PEP 8](https://www.python.org/dev/peps/pep-0008/). Other formatters are also supported.
```

> [Config reference](https://code.visualstudio.com/docs/python/settings-reference) - If you want to make your Python development on VS Code behave exactly like you want it to, view this reference guide. It explains options for the extension that can really improve your coding experience.

Reviewer 2: ```suggestion
[Config reference](https://code.visualstudio.com/docs/python/settings-reference) - If you want to configure your Python development on VS Code to behave exactly like you want it, view this reference guide. It explains options for the extensions which can really improve your coding experience.
```

---

### PyCharm

> PyCharm is an *Integrated Development Environment* built by JetBrains. It is specialized to work for Python and is commonly used among professionals. You can also extend it's features using plugins, but out-of-the-box it comes with a load of features pre-installed.

Reviewer 2: ```suggestion
PyCharm is an *Integrated Development Environment* built by JetBrains. It is specialized to work for Python and is commonly used among professionals. You can extend it plugins, but out-of-the-box it comes with a load of features already available.
```

#### Selecting the interpreter

Open your project, then navigate to `File` >> `Settings` >> `Project: ...` >> `Python Interpreter`. Click on the dropdown menu and select the environment you will be using. If the environment you would like to use is not in the list click on the `Show All...` button:

![Interpreter Selection Dropdown](https://raw.githubusercontent.com/exercism/python/main/docs/img/PyCharm-Config-InterpreterDropDown.png)

> From there click on the `+` button to add a new interpreter. Select the type of interpreter on the left, we suggest you either run a [conda]() or [virtualenv]() environment, but running the *system interpreter* works fine too. Once you selected your interpreter, press the `Okay` button.

Reviewer 2: ```suggestion
From there click on the `+` button to add a new interpreter. Select the type of interpreter on the left. We suggest you either run a [conda]() or [virtualenv]() environment, but running the *system interpreter* works fine, too. Once you selected your interpreter, press the `Okay` button.
```

![Add New Interpreter](https://raw.githubusercontent.com/exercism/python/main/docs/img/PyCharm-Config-InterpreterNew.png)

#### Other features

> Some other features that we won't cover in this guide, but are really useful for development are:

Reviewer 2: ```suggestion
Below are some other features that we won't cover in this guide, but are really useful for development.
```

[Running Tests](https://www.jetbrains.com/help/pycharm/pytest.html#run-pytest-test) - Running tests directly from a GUI in your window is really easy, but don't forget to look at the [pytest parameters](TESTS.md#extra-arguments) to set it up to your liking.

> [Debug Tools](https://www.jetbrains.com/help/pycharm/debugging-code.html) - Debugging in PyCharm can get really extensive, but can be really useful as well. Start here to learn how to properly debug using PyCharm.

Reviewer 2: I'm not sure "extensive" is the right word here.

Author: ```suggestion
[Debug Tools](https://www.jetbrains.com/help/pycharm/debugging-code.html) - Debugging in PyCharm is a great way to take advantage of it's tools. Start here to learn how to properly debug using PyCharm.
```

> [Debug Tools](https://www.jetbrains.com/help/pycharm/debugging-code.html) - Debugging in PyCharm is a great way to take advantage of it's tools. Start here to learn how to properly debug using PyCharm.

Reviewer 3: `Start here` should be linked to a url, right?

Author: Whoops, yep. Good catch.

Author: It refers to the link at the beginning of the `li`. I'll remove that sentence because it does not make any sense to have it there anyway.

```suggestion
[Debug Tools](https://www.jetbrains.com/help/pycharm/debugging-code.html) - Debugging in PyCharm is a great way to take advantage of its tools.
```

---

### Spyder IDE

> Spyder is a python IDE tailored for the scientific community. It packs some really good debugging utilities and it has an integrated *IPython terminal*. Out-of-the-box it does not come with support for *Unittests* and *Pytests*, but there is an official plugin available.

Reviewer 2: ```suggestion
Spyder is a python IDE tailored for the scientific community. It packs some really good debugging utilities and it has an integrated *IPython terminal*. Out-of-the-box it does not come with support for *Unittests* and *Pytests*, but there is an official plugin available for that.
```

#### Selecting the interpreter

To change the interpreter, go to `tools` >> `Preferences` >> `Python interpreter`. You can either select the global interpreter defined by Spyder or you can enter the path to your own Python environment.

![Spyder Python Interpreter Selection](https://raw.githubusercontent.com/exercism/python/main/docs/img/Spyder-Config-Interpreter.png)

Do not forget to click `Apply` once you selected your interpreter.

#### Other features

> [Spyder Unittest](https://github.com/spyder-ide/spyder-unittest/releases/latest) - If you want to have a built-in interface for you tests, install this plugin. Clicking the link will bring you to the latest release on GitHub.

Reviewer 2: ```suggestion
[Spyder Unittest](https://github.com/spyder-ide/spyder-unittest/releases/latest) - This plugin gives you a built-in interface for running your tests. This link will bring you to the latest release on GitHub.
```

> [Code analyzer](https://docs.spyder-ide.org/current/panes/pylint.html#using-the-code-analyzer) - An integrated tool for analyzing your code, detecting bad practices, and catching bugs before you run your program. You can also configure it to only catch [PEP 8](https://www.python.org/dev/peps/pep-0008/) violations.

Reviewer 2: ```suggestion
[Code analyzer](https://docs.spyder-ide.org/current/panes/pylint.html#using-the-code-analyzer) - This tool can analyze your code, detecting bad practices, and catching bugs before you run your program. You can also configure it to only show [PEP 8](https://www.python.org/dev/peps/pep-0008/) violations.
```

> [Debugging](https://docs.spyder-ide.org/current/panes/debugging.html) - This tool will help you debug your code, it allows you to set a `breakpoint` on a line. The debugging tool will then let you go through your python code and quickly notice where it goes wrong.

Reviewer 2: ```suggestion
[Debugging](https://docs.spyder-ide.org/current/panes/debugging.html) - This tool will help you debug your code. It allows you to set a `breakpoint` on a line. The debugging tool will then let you step through your python code line by line to find out where it goes wrong.
```

> [Variable Explorer](https://docs.spyder-ide.org/current/panes/variableexplorer.html) - This tool is really useful in combination with the debugger. It allows you to view all the active variables in your code and it even allows editing of the values inside those variables.

Reviewer 2: ```suggestion
[Variable Explorer](https://docs.spyder-ide.org/current/panes/variableexplorer.html) - This tool is really useful in combination with the debugger. It allows you to view all the active variables in your code and it even allows you to edit of the values inside those variables.
```

---

### Sublime text

> A text editor for coding, made by *Sublime HQ Pty Ltd*. It is similar to [VS Code](#visual-studio-code) in a lot of regards. Sublime text is almost fully customizable. It also has a built-in auto-complete engine and comes with Python syntax highlighting out-of-the-box.

Reviewer 2: ```suggestion
Sublime text is text editor for coding, made by *Sublime HQ Pty Ltd*. It is similar to [VS Code](#visual-studio-code) in many regards. Sublime text is very customizable. It comes with an auto-complete engine and with Python syntax highlighting out-of-the-box.
```

#### Notable extensions

> Sublime Text comes with a lot of tools already, but some of these tools could make your python development easier. Make sure you have *package control* installed, go to `Tools` >> `Install Package Control...`. To install packages open your command pallet and type in `Package Control: Install Package`, here you can enter the name of the packages you'd like to install.

Reviewer 2: ```suggestion
Sublime Text comes with a lot of tools. Here are some additional tools which can make your python development easier. Make sure you have *package control* installed in `Tools` >> `Install Package Control...`. To install packages, open your command pallet and type in `Package Control: Install Package` then enter the name of the packages you would like to install.
```

> [Terminus](https://packagecontrol.io/packages/Terminus) - A terminal for your sublime. It is fully featured and support for Windows operating systems. This package is almost a must-have for you to be able to run `pytest` scripts.

Reviewer 2: ```suggestion
[Terminus](https://packagecontrol.io/packages/Terminus) - A terminal for Sublime. Terminus is fully featured and has support for Windows operating systems. This package is a must-have for running `pytest` scripts.
```

> [SublimeLinter](https://packagecontrol.io/packages/SublimeLinter) - A linter for sublime. This tool automatically notifies you of bad-practice python code. Useful for when you have it set up to use PEP8.

Reviewer 2: ```suggestion
[SublimeLinter](https://packagecontrol.io/packages/SublimeLinter) - A linter for sublime. This tool notifies you about python code that does not follow best practices. We recommend you set it up to use PEP-8.
```

### JupyterLab

> Jupyter lab is an in-browser code editor that runs on your machine. You install it using a `pip` command in your *environment* and then automatically uses that environment. Jupyter is most well-known for being able to create coding notebooks. It is commonly used among data-scientists.

Reviewer 2: ```suggestion
Jupyter lab is an in-browser code editor that runs on your machine. You can install it using a `pip` command in your *environment* and Jupyter will use that environment. Jupyter is most well-known for its code "notebooks". It is commonly used among data-scientists.
```

#### Notable extensions

> The built-in tools for Jupyter Lab are good-enough for almost all exercises on *Exercism*. Most extensions made for Jupyter are made with data-science in mind. Explore extensions at your own pleasure.

Reviewer 2: ```suggestion
The built-in tools in Jupyter Lab are enough for the exercises on *Exercism*. Most extensions made for Jupyter are made with data-science in mind. Explore extensions at your own pleasure.
```

Author: "Enough", I am unsure if that is proper English. You say "Two days a week is enough" when you're talking about quantity, but do you say "The tool is enough for the job"? I am genuinely asking.

Reviewer: I typically put a qualifier for the second case, but not a hyphenate:  
"Is the tool good enough for the job?".  
or I might re-phrase and go with  "Can the tool handle the job?" 
or "Does the tool have the capabilities to handle the job?".

Slang-y-er forms of it might be:
"Is the tool up to the task?" 
or "Can it handle the task?".  

"Is it enough for the task." is also valid, but can feel ambiguous or  abrupt without the qualifier.  

And -- as a native speaker of AmEn, I would defer to what an international speaker feels is _less_  ambiguous without being too formal, or too UK - feeling.

Reviewer: and now that I really look at it, I might go with:

The built-in tools for Jupyter Lab are sufficient for almost all exercises on *Exercism*. Most extensions to Jupyter focus more on data-science applications and not general-purpose programming. Explore extensions at your own pleasure.

Author: > And -- as a native speaker of AmEn, I would defer to what an international speaker feels is less ambiguous without being too formal, or too UK - feeling.
Yep, I speak UK English, so... thank you!

I like this, thanks @BethanyG:
```suggestion
The built-in tools for Jupyter Lab are sufficient for almost all exercises on Exercism. Most extensions to Jupyter focus more on data-science applications and not general-purpose programming. Explore extensions at your own pleasure.
```
